### ISSUE TITLE
🐛 Unexpected results for floatDiffInRealDays (2.x) / diffInDays (3.x) on dst change

### ISSUE THREAD

**Issue by ederuiter** [briannesbitt/Carbon#2308]
Hello,

I encountered an issue with the following code:
```php
$start = new Carbon('2021-03-27 20:00:00', 'Europe/Amsterdam');
$end = new Carbon('2021-03-28 20:00:00', 'Europe/Amsterdam');
echo $start->floatDiffInRealDays($end), "\n";
echo $start->floatDiffInRealHours($end), "\n";
```
[2.x code example](https://try-carbon.herokuapp.com/?embed&theme=xcode&border=silver&options-color=rgba(120,120,120,0.5)&input=%24start%20%3D%20new%20Carbon(%272021-03-27%2020%3A00%3A00%27%2C%20%27Europe%2FAmsterdam%27)%3B%0A%24end%20%3D%20new%20Carbon(%272021-03-28%2020%3A00%3A00%27%2C%20%27Europe%2FAmsterdam%27)%3B%0Aecho%20%24start-%3EfloatDiffInRealDays(%24end)%2C%20%22%5Cn%22%3B%0Aecho%20%24start-%3EfloatDiffInRealHours(%24end)%2C%20%22%5Cn%22%3B%0A%0A%0A)

Carbon version: 2.46.0
PHP version: 7.4.16

I expected to get:

```
0.95833333333
23
```
But I actually get:

```
1.95833333333
23
```

This is related to a dst change on 2021-03-28 which causes that day to have 23 hours, this breaks the logic that adds the integer difference in days + the remaining hours ( `fmod($hoursDiff,  static::HOURS_PER_DAY) / static::HOURS_PER_DAY)` )

The same is the case for the 3.x branch ( [3.x code example](https://try-carbon.herokuapp.com/?embed&theme=xcode&border=silver&options-color=rgba(120,120,120,0.5)&input=%24start%20%3D%20new%20Carbon(%272021-03-27%2020%3A00%3A00%27%2C%20%27Europe%2FAmsterdam%27)%3B%0A%24end%20%3D%20new%20Carbon(%272021-03-28%2020%3A00%3A00%27%2C%20%27Europe%2FAmsterdam%27)%3B%0Aecho%20%24start-%3EdiffInDays(%24end)%2C%20%22%5Cn%22%3B%0Aecho%20%24start-%3EdiffInHours(%24end)%2C%20%22%5Cn%22%3B%0A%0A%0A) => the 3.x version did not seem to work for me, I keep getting `Update in progress, please retry in few minutes.`) where `->diffInHours()` returns 23 and the `$diff->format('%r%a')` returns 1

This bug results in weird experiences for the user as we determine the amount of rentalDays by calculating `ceil($start->floatDiffInRealDays($end))` 

```
'2021-03-27 20:00:00' - '2021-03-27 19:00:00' => 1 rental days (correct)
'2021-03-27 20:00:00' - '2021-03-27 20:00:00' => 2 rental days (incorrect)
'2021-03-27 20:00:00' - '2021-03-27 21:00:00' => 1 rental days (correct)
'2021-03-27 20:00:00' - '2021-03-27 22:00:00' => 2 rental days (correct)
```

Thanks!

**Comment by kylekatarnls** [comment]
Hello 👋

Sorry, this is a bug from PHP side (you would get some inconsistencies too with pure `DateInterval` object calculated from this diff).

No real work-around except doing all calculations in UTC:

```php
echo $start->utc()->floatDiffInRealDays($end->utc()), "\n";
echo $start->utc()->floatDiffInRealHours($end->utc()), "\n";
```

By the way city timezone like `Europe/Amsterdam` should be used only to parse an input of the user then to display the date to a given user you know use this timezone. But UTC should be your default system/DB/PHP/framework timezone setting and for any work you have to do back-end side.

See more in this article: https://kylekatarnls.medium.com/always-use-utc-dates-and-times-8a8200ca3164

**Comment by ederuiter** [comment]
Hi Kyle,

Thanks for the reply and possible workaround. 

In my opinion this boils down to a difference in definition, it seems `DateTIme::diff` considers the difference between `2021-03-27 20:00:00` and `2021-03-28 20:00:00` a complete day (although there are only 23 hours in it). Which seems sensible. 
The problems arise when combining that with a fractional part which is based on a definition that a complete day is always 24-hours. 
So I would be okay with these functions either returning `1` or `0.958333333` but not with `1.9583333333` 

Regarding utc usage, this is a bit more complex than just always use utc; see for example: https://javorszky.co.uk/2018/11/12/why-storing-future-times-in-timestamps-is-not-good/

For future rental appointments you would never want the times to shift if the dst rules are changed, thats why we store it as localtime + timezone in database.  And sure I can (and will) convert it to UTC for calculating the amount of rental days .. but in my opinion these are the kind of surprises that a library like Carbon wants to prevent.  

Kind regards,

  Eric

**Comment by kylekatarnls** [comment]
Hi,

Events would be fine to be stored as raw user input (as typed or picked in a calender component) + timezone of the user at that current moment. But it does not mean it's good to calculate difference with those timezones.

A first sad reason to not compute things from such timezoned dates is a long list of bugs unresolved for years from PHP (and datelib) side related to timezone and DST:
https://bugs.php.net/search.php?cmd=display&search_for=timezone&x=0&y=0

It would need a hack like `if (timezone has DST and diff contains a transition)` but this is obviously an heavy operation to search for transition in a given diff. I don't think it's worth it. Then each time we implemented a work-around for a DST bug, we had other unexpected behavior popping from it + conflicts with different PHP versions behaving differently.

That's one reason why 3.x will drop `diffInReal<Unit>` methods to just recommend instead `->utc()->diffIn<Unit>`.

Here is what I mean by UTC as the default. From a system point of view,
- The only correct way to represent a moment is an UTC datetime or timestamp (by the way, microsecond-precision is recommended), that one can be exchanged with other systems/DB/API, compared to other moments and new moments can be created from it.
- Date and time in timezones are translations of UTC moments (with loss due to DST, e.g. 1 time can represent 2 different moments during winter transitions) and uncertainty (as pointed by your example, we're not sure when exactly July 14th 2024 will start in Paris, maybe 00:00 GMT+1, maybe 00:00 GMT+2, depends if the country drops the DST yet or not) so those informations should not be considered as moments, not good for comparison, calculation, or to create new moments from it.

**Comment by kylekatarnls** [comment]
Will be fixed by #2312 which is basically converting dates to UTC before comparing them.
